crates/sui-indexer-alt-graphql/src/api/types/lookups.rs

```diff
@@ -23,77 +35,69 @@ use crate::{error::RpcError, pagination::Page};
 /// NOTE: for consistency, assume that lowerbounds are inclusive and upperbounds are exclusive.
 /// Bounds that do not follow this convention will be annotated explicitly (e.g. `lo_exclusive` or
 /// `hi_inclusive`).
-pub(crate) async fn tx_bounds<C>(
+pub(crate) async fn tx_bounds_query<'a>(
     ctx: &Context<'_>,
-    cp_bounds: &RangeInclusive<u64>,
-    global_tx_hi: u64,
-    page: &Page<C>,
-    f: fn(&C) -> u64,
-) -> Result<Range<u64>, RpcError> {
-    let pg_reader: &PgReader = ctx.data()?;
+    scope: &Scope,
+    filter: &impl TxBoundsFilter,
+    reader_lo: u64,
+    page: &Page<impl TxBoundsCursor>,
+) -> Result<Option<Query<'a>>, RpcError> {
+    if page.limit() == 0 {
+        return Ok(None);
+    }
+
+    let Some(checkpoint_viewed_at) = scope.checkpoint_viewed_at() else {
+        return Ok(None);
+    };
+
+    let Some(cp_bounds) = checkpoint_bounds(
```

Comment:
This also seems like it could/should be a function on or around the `TxBoundsFilter` trait.